src/zip.c

```diff
@@ -1038,3 +1074,395 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+    }
+    begin = i;
+    if (begin == entry_num) {
+      break;
+    }
+    while ((deleted_entry_index_array[i]) && (i < entry_num)) {
+      i++;
+    }
+    end = i;
+    int k = 0;
+    for (int j = end; j < entry_num; j++) {
+      MZ_ZIP_ARRAY_ELEMENT(&pState->m_central_dir_offsets, mz_uint32,
+                           begin + k) =
+          (mz_uint32)MZ_ZIP_ARRAY_ELEMENT(&pState->m_central_dir_offsets,
+                                          mz_uint32, j);
+      k++;
+    }
+    d_num += end - begin;
+  }
+
+  pState->m_central_dir_offsets.m_size =
+      sizeof(mz_uint32) * (entry_num - d_num);
+  return 0;
+}
+
+static int delete_entries(struct zip_t *zip,
+                          struct entry_mark *entry_mark_array, int entry_num) {
+  mz_bool *deleted_entry_flag_array =
+      (mz_bool *)calloc(entry_num, sizeof(mz_bool));
+  if (deleted_entry_flag_array == NULL) {
+    return -1;
+  }
+
+  mz_zip_internal_state *pState = zip->archive.m_pState;
+  zip->archive.m_zip_mode = MZ_ZIP_MODE_WRITING;
+
+  if (MZ_FSEEK64(pState->m_pFile, 0, SEEK_SET)) {
+    CLEANUP(deleted_entry_flag_array);
+    return -1;
+  }
+
+  mz_uint64 writen_num = 0;
+  mz_uint64 read_num = 0;
+  mz_uint64 deleted_length = 0;
+  mz_uint64 move_length = 0;
+  int i = 0;
+  int deleted_entry_num = 0;
+  while (i < entry_num) {
+    while ((entry_mark_array[i].type == KEEP) && (i < entry_num)) {
+      writen_num += entry_mark_array[i].lf_length;
+      read_num = writen_num;
+      i++;
+    }
+
+    while ((entry_mark_array[i].type == DELETE) && (i < entry_num)) {
+      deleted_entry_flag_array[i] = MZ_TRUE;
+      read_num += entry_mark_array[i].lf_length;
+      deleted_length += entry_mark_array[i].lf_length;
+      i++;
+      deleted_entry_num++;
+    }
+
+    while ((entry_mark_array[i].type == MOVE) && (i < entry_num)) {
+      move_length += entry_mark_array[i].lf_length;
+      mz_uint8 *p = &MZ_ZIP_ARRAY_ELEMENT(
+          &pState->m_central_dir, mz_uint8,
+          MZ_ZIP_ARRAY_ELEMENT(&pState->m_central_dir_offsets, mz_uint32, i));
+      if (!p) {
+        CLEANUP(deleted_entry_flag_array);
+        return -1;
+      }
+      mz_uint32 offset = MZ_READ_LE32(p + MZ_ZIP_CDH_LOCAL_HEADER_OFS);
+      offset -= (mz_uint32)deleted_length;
+      MZ_WRITE_LE32(p + MZ_ZIP_CDH_LOCAL_HEADER_OFS, offset);
+      i++;
+    }
+    if (move_files(pState->m_pFile, writen_num, read_num, move_length) !=
+        (mz_int64)move_length) {
+      CLEANUP(deleted_entry_flag_array);
+      return -1;
+    }
+    writen_num += move_length;
+    read_num += move_length;
+  }
+
+  zip->archive.m_archive_size -= deleted_length;
+  zip->archive.m_total_files = entry_num - deleted_entry_num;
+
+  delete_central_dir_entrys(pState, deleted_entry_flag_array, entry_num);
+  CLEANUP(deleted_entry_flag_array);
+
+  return deleted_entry_num;
+}
+
+int zip_entries_delete(struct zip_t *zip, char *entries[], const int num) {
+  if (zip == NULL || entries == NULL) {
```

Comment:
I think if `entries` is NULL we can do nothing, just return without any error code. The problem is if `entries` is NULL but `num > 0`

Reply:
OK,  I'll fix it.